### ISSUE TITLE
Error:ORA-01036: illegal variable name/number

### ISSUE THREAD

**Issue by zipiii** [jcachat/EntityFramework.DynamicFilters#114]
I am trying to implement db context global filtering using EntityFramework.DynamicFilters
 I am using EF6, Oracle 11g and dotConnect for Oracle 9.2.187.
 Unfortunatelly i got an error in case of using Dynamic filter：ORA-01036: illegal variable name/number.

I am attaching simple project where the situation can be replayed.
https://pan.baidu.com/s/1slk8GK9
 or
https://github.com/zipiii/EFDynamicFilterDemoOfDevart

To run the project please modify connection string, create oracle db, create schema abp1 and run Update-Database migration command and run it.

When using Oracle Devart connection, EntityFramework DynamicFilters can only be used under the 1.4.9 version, if use 1.4.10 above and got an error:ORA-01036: illegal variable name/number.

http://forums.devart.com/viewtopic.php?f=1&t=35458&p=123605#p123605

Please help me!

**Comment by jcachat** [comment]
I do not have a license for dotConnect so I can't test this.  Officially, I only support the Official Oracle ODP.NET, Managed Entity Framework Driver as described on the main page of this project.  So I cannot guarantee what will or won't work when using a different driver.

But...I read through the discussion on the devart forum.  They found that the problem was being caused by a parameter being manually removed.  That was done for issue #53 so you can read why it was done there.  There is no way for me to also remove the parameter from the parameter collection because that can cause other issues with some other queries and some other drivers consider that collection to be read-only.

A possible workaround for that would be to disable how DynamicFilters adds the "disable filter" clause to the query in the first place - which would prevent the filter from being added beforehand.  Instructions are at the bottom of the issue referenced above (search for "PreventAllDisabledFilterConditions").  Just be aware that if you do that, you will not have the ability to later enable/disable the filter.

The only other thing I could do would be to add an option to turn off removing the parameter during the sql interception (to essentially disable the change made for that issue).  Or to do it for specific databases/drivers (there is another open issue for SQL Server CE - #81 - so I could fix both of these at once).  But I won't be able to get to that for another week or 2.  So try the suggestion above - at least as a temporary workaround - and report back on whether or not that solves the problem.

I am willing to make a change on my side to try to get this to work.  But, not to point fingers...the other option is for Devart to change their driver to behave like all of the other .Net EF drivers (including the official Oracle driver) which do not have any issue with this at all.

**Comment by zipiii** [comment]
Hi, jcachat.
Thank you very much.
I already gave you reply message to deavrt, waiting for their reply.
Will immediately tell you after I have progress.

**Comment by jcachat** [comment]
Ok, but what I posted was mostly for you to try to solve the problem on your side - without requiring a change from DevArt.  If they can change their EF driver to not throw the error, that's great.  But please still try adding a call to "PreventAllDisabledFilterConditions" (in your OnModelCreating after you have defined all of your filters) and verify if that at least works around the problem.  If so, I will add an option to turn off removing that parameter from the intercepted sql.  I believe I need to do that for SQL CE anyway.

**Comment by jcachat** [comment]
I pushed out a new build (NuGet package 2.8) that should prevent removing the "filter is disabled" clauses during sql interception when using the DevArt Oracle driver.  Since I don't have a license to that provider, I had to guess at the provider name based on the same project you uploaded.  So please try that update and let me know how it works.

**Comment by zipiii** [comment]
Hi jcachat, I try this update, then process and results are as follows:
1. I update package 1.4.9 to 2.8.0.
2. I add the code in DbContext.OnModelCreating:
   `modelBuilder.Filter("SoftDelete", (ISoftDelete d) => d.IsDeleted, true);
    modelBuilder.PreventDisabledFilterConditions("SoftDelete");`
3. In tester run:
  ```
     var context = new ExampleContext();
     //context.DisableFilter("SoftDelete");
     context.EnableFilter("SoftDelete");
     Query(context, "homer");
```     
It is ok, 
     but I add context.DisableFilter("SoftDelete") before context.EnableFilter("SoftDelete") then it is failure:
   ```
var context = new ExampleContext();
   context.DisableFilter("SoftDelete");
     context.EnableFilter("SoftDelete");
     Query(context, "homer");
```
The exception message is "Enable/Disable filters conditions have been turned off via PreventDisabledFilterConditions!”

hint: I upload devart oracle driver to http://pan.baidu.com/s/1o87aPEa, you can download it and try it.
thank you！

**Comment by jcachat** [comment]
With the 2.8 update you don't need PreventDisabledFilterConditions().  I was suggesting to try that before I made the change but since I didn't hear a response back from you, I went ahead and made the change anyway.

Using PreventDisabledFilterConditions() causes the library to not add the conditions for disabling the filters.  That is why you get that error.

**Comment by jcachat** [comment]
You should also remove that oracle driver that you uploaded to a public site.  I don't think DevArt will like that.

**Comment by zipiii** [comment]
I'm removed the download site.

Need to add any particular code If use 2.8 update ?

**Comment by zipiii** [comment]
I used 2.8 update, if i not change the code, then it is error: ORA-01036: illegal variable name/number.
The code is as follows:
```
var context = new ExampleContext();
   context.EnableFilter("SoftDelete");
   Query(context, "homer");
```

The devart provide the Trial version: https://www.devart.com/dotconnect/oracle/download.html.
Please try it, thank you!

**Comment by jcachat** [comment]
I am not able to get the DevArt provider to connect to my oracle instance.  I probably just don't know enough about it.  But, I was able to tell that I had the case wrong on the string comparison - I was looking for "DevArt" when the provider is named "Devart".  I pushed out another update - 2.8.1.  Try that one.

If that doesn't work, I'm going to need you to download the source and do some debugging.
